core/blockchain/impl/block_tree_impl.cpp

```diff
@@ -1231,78 +1252,71 @@ namespace kagome::blockchain {
-                                             -> outcome::result<
-                                                 primitives::BlockInfo> {
-      OUTCOME_TRY(target_header, p.header_repo_->getBlockHeader(target_hash));
-      if (max_number.has_value() && target_header.number > max_number.value()) {
-        return BlockTreeError::TARGET_IS_PAST_MAX;
-      }
-      OUTCOME_TRY(canon_hash,
-                  p.header_repo_->getHashByNumber(target_header.number));
-      // if a max number is given we try to fetch the block at the
-      // given depth, if it doesn't exist or `max_number` is not
-      // provided, we continue to search from all leaves below.
-      if (canon_hash == target_hash) {
-        if (max_number.has_value()) {
-          OUTCOME_TRY(hash_opt, p.storage_->getBlockHash(max_number.value()));
-          if (hash_opt.has_value()) {
-            return primitives::BlockInfo{max_number.value(), hash_opt.value()};
+      const primitives::BlockHash &target_hash) const {
+    return block_tree_data_.sharedAccess(
+        [&](const auto &p) -> outcome::result<primitives::BlockInfo> {
+          auto &root = p.tree_->getRoot();
+
+          auto target = root.findByHash(target_hash);
+
+          // If target has not found in block tree (in memory),
+          // it means block finalized or discarded
+          if (not target) {
+            OUTCOME_TRY(target_number,
+                        p.header_repo_->getNumberByHash(target_hash));
+
+            OUTCOME_TRY(canon_hash,
+                        p.header_repo_->getHashByNumber(target_number));
+
+            if (canon_hash != target_hash) {
+              return BlockTreeError::BLOCK_ON_DEAD_END;
+            }
+
+            target = root.findByHash(root.block_hash);
           }
-        }
-      } else {
-        OUTCOME_TRY(
-            last_finalized,
-            p.header_repo_->getNumberByHash(getLastFinalizedNoLock(p).hash));
-        if (last_finalized >= target_header.number) {
-          return BlockTreeError::BLOCK_ON_DEAD_END;
-        }
-      }
-      for (auto &leaf_hash : getLeavesSortedNoLock(p)) {
-        auto current_hash = leaf_hash;
-        auto best_hash = current_hash;
-        if (max_number.has_value()) {
-          OUTCOME_TRY(
-              hash,
-              walkBackUntilLessNoLock(p, current_hash, max_number.value()));
-          best_hash = hash;
-          current_hash = hash;
-        }
-        OUTCOME_TRY(best_header, p.header_repo_->getBlockHeader(best_hash));
-        primitives::BlockNumber current_block_number{};
-        do {
-          OUTCOME_TRY(current_header,
-                      p.header_repo_->getBlockHeader(current_hash));
-          if (current_hash == target_hash) {
-            return primitives::BlockInfo{best_header.number, best_hash};
+
+          auto metadata = p.tree_->getMetadata();
+
+          std::multiset<std::shared_ptr<TreeNode>> candidates{};
```

Comment:
```suggestion
          std::set<std::shared_ptr<TreeNode>, DerefLess> candidates;
```

Reply:
`multiset -> set` - ok
`DerefLess` - does not exists in current branch

Reply:
`DerefLess(sptr l, sptr r) { *l < *r }`